libraries/AP_InertialSensor/AP_InertialSensor_tempcal.cpp

```diff
@@ -0,0 +1,510 @@
+  reference. This makes the SITL data independent of TEMP_REFERENCE
+  and prevents an abrupt change at the endpoints
+ */
+void AP_InertialSensor::TCal::sitl_apply_accel(float temperature, Vector3f &accel) const
+{
+    const float tmid = 0.5*(temp_max+temp_min);
+    accel += polynomial_eval(temperature - tmid, accel_coeff);
+}
+
+void AP_InertialSensor::TCal::sitl_apply_gyro(float temperature, Vector3f &gyro) const
+{
+    const float tmid = 0.5*(temp_max+temp_min);
+    gyro += polynomial_eval(temperature - tmid, gyro_coeff);
+}
+
+AP_InertialSensor::TCal::Learn::Learn(TCal &_tcal, float _start_temp) :
+    start_temp(_start_temp),
+    tcal(_tcal)
+{
+    reset(_start_temp);
+}
+
+/*
+  update polyfit with new sample
+ */
+void AP_InertialSensor::TCal::Learn::add_sample(const Vector3f &sample, float temperature, struct LearnState &st)
+{
+    temperature = st.temp_filter.apply(temperature);
+
+    st.sum += sample;
+    st.sum_count++;
+
+    uint32_t now = AP_HAL::millis();
+
+    if (st.sum_count < 100 ||
+        temperature - st.last_temp < 0.5) {
+        // check for timeout
+        if (st.last_sample_ms != 0 &&
+            temperature - start_temp >= TEMP_RANGE_MIN &&
+            now - st.last_sample_ms > CAL_TIMEOUT_MS) {
+            // we have timed out, finish up now
+            finish_calibration(st.last_temp);
+        }
+        // wait for more data
+        return;
+    }
+
+    st.sum /= st.sum_count;
+
+    const uint8_t si = &st - &state[0];
+
+    const float T = (temperature + st.last_temp) * 0.5;
+
+    if (si == 0) {
+        // we use the first accel sample as the zero baseline
+        if (accel_start.is_zero()) {
+            accel_start = st.sum;
+            start_temp = T;
+        }
+        st.sum -= accel_start;
+    }
+
+    const float tdiff = T - TEMP_REFERENCE;
+
+    AP::logger().Write("TCLR", "TimeUS,I,SType,Temp,X,Y,Z,NSamp",
+                       "s#------",
+                       "F000000-",
+                       "QBBffffI",
+                       AP_HAL::micros64(),
+                       instance(),
+                       si,
+                       T,
+                       st.sum.x, st.sum.y, st.sum.z,
+                       st.sum_count);
+    
+    
+    st.pfit.update(tdiff, st.sum);
+
+    st.sum.zero();
+    st.sum_count = 0;
+    st.last_temp = temperature;
+    st.last_sample_ms = now;
+
+    if (temperature - start_temp >= TEMP_RANGE_MIN) {
+        if (temperature >= start_tmax) {
+            // we've reached the target temperature
+            finish_calibration(temperature);
+        } else {
+            // save partial calibration, so if user stops the cal part
+            // way then they still have a useful calibration
+            save_calibration(st.last_temp);
+        }
+    }
+}
+
+/*
+  update accel temperature compensation learning
+ */
+void AP_InertialSensor::TCal::update_accel_learning(const Vector3f &accel, float temperature)
```

Comment:
We're mixing implementations for functions in different namespaces  through here.

Reply:
not sure what you mean? This file implements a class and a subclass